prime-router/src/main/kotlin/tokens/TokenAuthentication.kt

```diff
@@ -0,0 +1,220 @@
+            if (bearerComponent == null) {
+                logger.error("Missing Authorization header.  Unauthorized.")
+                return null
+            }
+            val accessToken = bearerComponent.split(" ")[1]
+            if (accessToken.isEmpty()) {
+                logger.error("Request has Authorization header but no token.  Unauthorized")
+                return null
+            }
+// for testing only           return checkAccessToken(accessToken, desiredScope, GetStaticSecret())
+            return checkAccessToken(accessToken, desiredScope, FindReportStreamSecretInVault())
+        }
+
+        /**
+         * lookup is a call back to get the ReportStream secret used to sign the token.  Using a callback
+         * makes this easy to test - can pass in a static test secret
+         * This does not need to be a public/private key.
+         */
+        fun checkAccessToken(accessToken: String, desiredScope: String, lookup: ReportStreamSecretFinder): Claims? {
+            try {
+                val secret = lookup.getReportStreamTokenSigningSecret()
+                // Check the signature.  Throws JwtException on problems.
+                val jws = Jwts.parserBuilder()
+                    .setSigningKey(secret)
+                    .build()
+                    .parseClaimsJws(accessToken)
+                if (jws.body == null) {
+                    logger.error("AccessToken check failed - no claims.  Unauthorized.")
+                    return null
+                }
+                if (isExpiredToken(jws.body.expiration)) {
+                    logger.error("AccessToken expired.  Unauthorized.")
+                    return null
+                }
+                val scope = jws.body["scope"] as? String
+                if (scope == null) {
+                    logger.error("Missing scope claim.  Unauthorized.")
+                    return null
+                }
+                if (!Scope.scopeListContainsScope(scope, desiredScope)) {
+                    logger.error("Signed key has scope $scope, but requests $desiredScope.  Unauthorized")
+                    return null
+                }
+                return jws.body
+            } catch (ex: JwtException) {
+                logger.error("Unauthorized: $ex")
+                return null
+            } catch (exc: RuntimeException) {
+                logger.error("Unauthorized: $exc")
+                return null
+            }
+        }
+
+    companion object: Logging {
+        fun isExpiredToken(exp: Date): Boolean {
+            return (Date().after(exp))   // no need to include clock skew, since we generated token ourselves
+        }
+    }
+}
+
+/**
+ * Defined per the FHIR standard
+ *    https://hl7.org/fhir/uv/bulkdata/authorization/index.html
+ */
+// todo get this to work:  @JsonNaming(PropertyNamingStrategy.SnakeCaseStrategy.class)
+data class AccessToken(
+    @JsonProperty("access_token")
```

Comment:
I'm wondering if we should generate a unique GUID for each AccessToken. That way can track its issuance and use in our logs. I would put this as the subject of the token.

Reply:
Done.   Added the GUID and logging, and its in the Action history table too.